develop/items/custom-data-components.md

```diff
@@ -0,0 +1,267 @@
+
+```java
+int clickCount = stack.getOrDefault(ModItems.CLICK_COUNT_COMPONENT, 0);
+```
+
+As you can see this method takes two arguments, our component type like before and a default value to return if the component is not present.
+
+If you implement either of these solutions and hover over the item without the component you should see that it displays "Used 0 times" and no longer crashes the game.
+
+![Tooltip showing "Used 0 times"](/assets/develop/items/custom_component_2.png)
+
+## Updating the component value {#setting-component-value}
+
+Now let's try updating our component value. We're going to increase the click count each time we use our Lightning Stick item. To change the value of a component on an `ItemStack` we use the `set()` method like so:
+
+```java
+stack.set(ModItems.CLICK_COUNT_COMPONENT, newValue);
+```
+
+This takes our component type and the value we want to set it to. In this case it will be our new click count.
+
+Let's add a few lines to our `use()` method to read the old click count, increase it by one and then set the updated click count.
+
+```java
+@Override
+public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {
+    // spawn lightning
+    ...
+
+    // Update the click count
+    ItemStack stack = user.getStackInHand(hand);
+    int clickCount = stack.getOrDefault(ModItems.CLICK_COUNT_COMPONENT, 0);
+    stack.set(ModItems.CLICK_COUNT_COMPONENT, ++clickCount);
+
+    return TypedActionResult.success(stack);
+}
+```
+
+Now try starting the game and right clicking with the Lightning Stick. Count how many times you right click it. If you open up your inventory and look at the item again you should see that the usage number has gone up by the amount you clicked it.
+
+![Tooltip showing "Used 8 times"](/assets/develop/items/custom_component_3.png)
+
+## Removing the component value {#removing-component-value}
+
+You can also remove a component from your `ItemStack` if it is no longer needed. This is done by using the `remove()` method which takes your component type.
+
+```java
+stack.remove(ModItems.CLICK_COUNT_COMPONENT);
+```
+
+This method also returns the value of the component before being removed, so you can also use it like so:
+
+```java
+int oldCount = stack.remove(ModItems.CLICK_COUNT_COMPONENT);
+```
+
+## Advanced Data Components (maps) {#advanced-data-components}
+
+You may need to store multiple attributes in a single component. As a vanilla example, the `minecraft:food` component stores several values related to food, such as `nutrition`, `saturation`, `eat_seconds` and more. These components are referred to by the game as "map components".
```

Comment:
> These components are referred to by the game as "map components"

Are they? Where? If you came up with that name, then I wouldn't say it's very good.

Reply:
This is based on the error message in the missing components which says `No key temperature in MapLike`. But I couldn't find any official name for them in the announcement or docs. Would calling them "composite components" be a bit more fitting maybe?